src/profile-logic/transforms.js

```diff
@@ -98,190 +99,222 @@ const SHORT_KEY_TO_TRANSFORM: { [string]: TransformType } = {};
-        const callNodePath = stringToUintArray(serializedCallNodePath);
-        const inverted = Boolean(invertedRaw);
-        // Flow requires a switch because it can't deduce the type string correctly.
-        switch (type) {
-          case 'focus-subtree':
-            transforms.push({
-              type: 'focus-subtree',
-              implementation,
-              callNodePath,
-              inverted,
-            });
-            break;
-          case 'merge-call-node':
-            transforms.push({
-              type: 'merge-call-node',
-              implementation,
-              callNodePath,
-            });
-            break;
-          default:
-            throw new Error('Unmatched transform.');
-        }
 
-        break;
+          break;
+        }
+        default:
+          throw assertExhaustiveCheck(type);
       }
-      default:
-        throw assertExhaustiveCheck(type);
-    }
-  });
-
-  return transforms;
+    });
+  }
+  return transformStacksPerThread;
 }
 
-export function stringifyTransforms(transforms: TransformStack = []): string {
-  return transforms
-    .map(transform => {
-      const shortKey = TRANSFORM_TO_SHORT_KEY[transform.type];
-      if (!shortKey) {
-        throw new Error(
-          'Expected to be able to convert a transform into its short key.'
-        );
-      }
-      // This switch breaks down each transform into shared groups of what data
-      // they need, as defined in src/types/transforms.js. For instance some transforms
-      // need only a funcIndex, while some care about the current implemention, or
-      // other pieces of data.
-      switch (transform.type) {
-        case 'merge-function':
-        case 'drop-function':
-        case 'collapse-function-subtree':
-        case 'focus-function':
-          return `${shortKey}-${transform.funcIndex}`;
-        case 'collapse-resource':
-          return `${shortKey}-${transform.implementation}-${transform.resourceIndex}-${transform.collapsedFuncIndex}`;
-        case 'collapse-direct-recursion':
-          return `${shortKey}-${transform.implementation}-${transform.funcIndex}`;
-        case 'focus-subtree':
-        case 'merge-call-node': {
-          let string = [
-            shortKey,
-            transform.implementation,
-            uintArrayToString(transform.callNodePath),
-          ].join('-');
-          if (transform.inverted) {
-            string += '-i';
+/**
+ * Each transform in the stack is separated by a ",".
+ * Each thread is separated by a ";".
+ * The thread order matches the selected thread indexes order.
+ */
+export function stringifyTransforms(
+  selectedThreads: Set<ThreadIndex>,
+  transformStacksPerThread: TransformStacksPerThread
+): string {
+  // The iterator for the Set<ThreadIndex> will give the threads in the same order.
+  // The order itself is arbitrary, but it should be consistent across the stringify
+  // calls.
+  return [...selectedThreads]
```

Comment:
I wonder if we can remove this intermediate array by getting the iterator directly. Don't think it's such a big deal since this set will be fairly small. We can directly use `for..of` or use `forEach` on the map. Although, I believe current `.map` is better for readability and telling the people that we map the values directly 1-1. What do you think?

Reply:
I just profiled this a bit, and it was only 1 sample in a bunch of profiler interaction, so I'm not concerned. The logic is a bit tricker without being able to "join". An optimization would be to memoize, as the result here is mostly just reused, although I don't think it'll be a big win.